clients/src/main/java/org/apache/kafka/clients/producer/ProducerRecord.java

```diff
@@ -56,6 +59,8 @@
     public ProducerRecord(String topic, Integer partition, Long timestamp, K key, V value) {
         if (topic == null)
             throw new IllegalArgumentException("Topic cannot be null");
+        if (!TopicValidator.isValidateTopicName(topic))
+            throw new InvalidTopicException(topic + " is not a valid topic name.");
```

Comment:
The logic for a valid topic name is actually a little more complicated than this. The server side validation exists in `kafka.common.Topic.scala`.

We also shouldn't duplicate the logic. So if we move it to the clients package, removing the Scala version and replacing its usages would be good too.

Also #898 (KAFKA-3219) may change this logic a little bit too.